### PR TITLE
use atomic store provider in AMM

### PR DESCRIPTION
**Pull request by turadg**
## Description

For https://github.com/Agoric/agoric-sdk/pull/5721 I had to make an atomic provider helper to avoid a race condition. @dtribble pointed out this is similar to what https://github.com/Agoric/agoric-sdk/pull/5596 did.

#5721 landed. This PR refactors the AMM's `makeAddIssuer` to use the new `makeAtomicProvider` helper. It improves the helper's types, docs and tests.

### Security Considerations

--

### Documentation Considerations

Post in chat about new helper

### Testing Considerations

Expanded test coverage on helper, to have confidence in `finish` option.

### CODE REVIEW

File: packages/store/src/stores/store-utils.js [hunk 1]
```diff
@@ -106,3 +106,69 @@ export const provide = (mapStore, key, makeValue) => {
...
+ * possible for termination to happen after the make completes and before it
+ * reaches durable storage.
+ *
+ * @template K
+ * @template V
+ * @param {MapStore<K, V>} durableStore
+ */
+export const makeAtomicProvider = durableStore => {
+  /** @type {Map<K, Promise<V>>} */
+  const pending = new Map();
```
Comment (turadg) [inline packages/store/src/stores/store-utils.js:127]: should this be a WeakMap? a WeakStore?

File: packages/store/src/stores/store-utils.js [hunk 2]
```diff
@@ -106,3 +106,69 @@ export const provide = (mapStore, key, makeValue) => {
...
+   * If there already is one, return that. Otherwise,
+   * call `makeValue(key)`, remember it as the value for
+   * that key, and return it.
+   *
+   * @param {K} key
+   * @param {(key: K) => Promise<V>} makeValue
+   * @param {(key: K, value: V) => Promise<void>} [finishValue]
+   * @returns {Promise<V>}
+   */
+  const provideAsync = (key, makeValue, finishValue) => {
```
Comment (Chris-Hibbert) [inline packages/store/src/stores/store-utils.js:141]: finishValue isn't described in the summary above, and isn't tested.   I'd like to see both.

File: packages/store/src/stores/store-utils.js (hunk 1 above)
Comment (Chris-Hibbert) [reply packages/store/src/stores/store-utils.js:127]: If the provider crashes, the store will go away. If the provider doesn't crash, each key will be removed as it resolves.  I don't think there's a need for a weak collection.

File: packages/run-protocol/src/vpool-xyk-amm/addPool.js [hunk 3]
```diff
@@ -44,65 +41,54 @@ export const makeAddIssuer = (
...
+      // defer lookup until necessary. more aligned with governed
+      // param we expect this to be eventually.
+      const addIssuerToReserve = getAddIssuerToReserve();
+
+      // tell the reserve about this brand, which it will validate by
+      // calling back to AMM for the issuer
+      return addIssuerToReserve(brand);
+    };
+
+    const liquidityMint = await brandToLiquidityMintProvider.provideAsync(
```
Comment (Chris-Hibbert) [inline packages/run-protocol/src/vpool-xyk-amm/addPool.js]: Be consistent about `// AWAIT /////////` markers within a function! There's one on the `await` above, so there must be one here as well.

File: packages/run-protocol/src/vpool-xyk-amm/addPool.js [hunk 4]
```diff
@@ -44,65 +41,54 @@ export const makeAddIssuer = (
...
+          return mint;
+        })
+        .catch(e => {
+          console.error(
+            X`Failure Saving issuer ${secondaryIssuer}. Not added to Reserve`,
+          );
+          throw e;
+        });
+    };
+    /** @type {(brand: Brand) => Promise<void>} */
```
Comment (Chris-Hibbert) [inline packages/run-protocol/src/vpool-xyk-amm/addPool.js:73]: I'd like to see a blank line between `makeLiquidityMint` and `finish`.  The break was hard for me to parse visually.

**Review by Chris-Hibbert — COMMENTED** [review]
I think the change is clarifying. Thanks.

I'll approve once tests are added.

File: packages/run-protocol/src/vpool-xyk-amm/addPool.js (hunk 3 above)
Comment (turadg) [reply packages/run-protocol/src/vpool-xyk-amm/addPool.js]: I can't find any documentation about when and when not to use `// AWAIT`. Does it just mean that there's an `await` token on the next line? If so, that seems like a code editing concern, not source code.

If we are going to be consistent, shouldn't we be consistent across the repo?

I'll make the change once I can read documentation on when it's supposed to be used. I'll make the documentation if necessary.

Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/addPool.js]: > I can't find any documentation about when and when not to use // AWAIT. Does it just mean that there's an await token on the next line? If so, that seems like a code editing concern, not source code.

Usually on the previous line.    We use it to raise awareness in functions with delicate timing/interaction. That's why I'm asking for consistency within a function.

> If we are going to be consistent, shouldn't we be consistent across the repo?

We've found it to be distracting when overused.

> I'll make the change once I can read documentation on when it's supposed to be used. I'll make the documentation if necessary.

I'd talk with @dtribble

Comment (turadg) [reply packages/run-protocol/src/vpool-xyk-amm/addPool.js]: For now I removed the `await`.

File: packages/run-protocol/src/vpool-xyk-amm/multipoolMarketMaker.js [hunk 5]
```diff
@@ -166,15 +166,13 @@ const start = async (zcf, privateArgs) => {
   const isSecondary = secondaryBrandToPool.has;
 
   // The liquidityBrand has to exist to allow the addPool Offer to specify want
-  /** @type {WeakStore<Brand,ZCFMint<'nat'>>} */
+  /** @type {WeakMapStore<Brand,ZCFMint<'nat'>>} */
```
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/multipoolMarketMaker.js:169]: `WeakStore` type is deprecated, says to use `WeakMapStore`

**Review by Chris-Hibbert — APPROVED** [review]

**Pull request merged**